text/add-bf16-f64f64-and-f80-type.md

```diff
@@ -0,0 +1,121 @@
+- Feature Name: `add-bf16-f64f64-and-f80-type`
+- Start Date: 2023-7-10
+- RFC PR: [rust-lang/rfcs#3456](https://github.com/rust-lang/rfcs/pull/3456)
+- Rust Issue: [rust-lang/rust#2629](https://github.com/rust-lang/rfcs/issues/2629)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes new floating point types to enhance FFI with specific targets:
+
+- `bf16` as builtin type for the 'brain floating point' format, widely used in machine learning, different from the IEEE 754 standard `binary16` representation
+- `f64f64` in `core::arch` for the legacy extended float format used in the PowerPC architecture
+- `f80` in `core::arch` for the extended float format used in the x86 and x86_64 architectures
+
+Also, this proposal introduces `c_longdouble` in `core::ffi` to represent the correct format for 'long double' in C.
+
+# Motivation
+[motivation]: #motivation
+
+The types listed above may be widely used in existing native code, but are not available on all targets. Their underlying representations are quite different from 16-bit and 128-bit binary floating format defined in IEEE 754.
+
+In respective targets (namely PowerPC and x86), the target-specific extended types are referenced by `long double`, which makes `long double` ambiguous in the context of FFI. Thus defining `c_longdouble` should help interoperating with C code using the `long double` type.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`bf16` is available on all targets. The operators and constants defined for `f32` are also available for `bf16`.
+
+For `f64f64` and `f80`, their availability is limited to the following targets, but this may change over time:
+
+- `f64f64` is supported on `powerpc-*` and `powerpc64(le)-*`, available in `core::arch::{powerpc, powerpc64}`
+- `f80` is supported on `i[356]86-*` and `x86_64-*`, available in `core::arch::{x86, x86_64}`
+
+The operators and constants defined for `f32` and `f64` are available for `f64f64` and `f80` in their respective arch-specific modules.
+
+All proposed types do not have literal representation. Instead, they can be converted to or from IEEE 754 compliant types.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `bf16` type
+
+`bf16` consists of 1 sign bit, 8 bits of exponent, 7 bits of mantissa. Some ARM, AArch64, x86 and x86_64 targets support `bf16` operations natively. For other targets, they will be promoted into `f32` before computation and truncated back into `bf16`.
```

Comment:
Is that equivalent to whatever the IEEE semantics of bf16 are, if such a thing exists?

Reply:
Reading other comments below it seems like this is actually an incorrect emulation. So it will be the case, for the first time in Rust history, that primitive operations such as multiplying two elements of `bf16` have target-dependent behavior. That's a major downside of the RFC and needs to be discussed and justified more explicitly. The RFC should also state explicitly what is *guaranteed* to be true about `bf16` arithmetic on all targets -- that's needed e.g. for unsafe code authors to know what they can rely on in terms of soundness.

Reply:
It should be possible to emulate correct rounding semantics, so this frankly seems like a bug in those softfp impls.

It does require temporarily switching to RTZ mode, and then you can truncate the result with RTN-ties-even.

Reply:
Or maybe alternatively the answer for all these float types is -- the resulting bit patterns are entirely unspecified and not guaranteed to be portable in any way.

But even then we need to document how deterministic they are. Does passing the exact same inputs to the operation multiple times during a program execution always definitely produce the exact same outputs, on all targets and optimization levels? For regular floats, the answer turns out to be "only when there are no NaNs" -- that's what https://github.com/rust-lang/rfcs/pull/3514 is all about.

Reply:
iirc, you can do a single add/sub/mul/div/sqrt `bf16` operation by promoting to `f32`, doing the operation in `f32`, and then rounding back to `bf16` using the same rounding mode, *not* truncating. That is assuming, of course, that `bf16` actually meets the conditions which are iirc something like having `bf16`'s mantissa bit count be less than half of `f32`'s mantissa bit count minus 1 or 2.

this is like how you can do that with `f32` and `f64`, which is how you can do `f32` operations in JavaScript by using `Math.fround`.

Reply:
That's true for the regular IEEE formats, yes -- it was proven [here](https://hal.science/hal-01091186/document).

But I don't know if bf16 is enough like an IEEE format to make that theorem apply.

Reply:
> That's true for the regular IEEE formats, yes -- it was proven [here](https://hal.science/hal-01091186/document).
> 
> But I don't know if bf16 is enough like an IEEE format to make that theorem apply.

it is, `bf16` is just `f16` with a few more exponent bits and a few less mantissa bits, everything else is the same.

Reply:
`bf16` is just `f32` with the lower 16 mantissa bits dropped. As `f64` values that can be rounded to `f32` are effectively `f32` values with 29 extra mantissa bits, there would be no difference here.

Reply:
> Also, does LLVM when it compiles bf16 to f32 guarantee to do the rounding back to `bf16` after _each and every operation_, never doing more than one operation "at once" in `f32` mode?

that I don't know, but I hope LLVM at least tries to be correct in non-fast-math mode

Reply:
> that I don't know, but I hope LLVM at least tries to be correct in non-fast-math mode

That should definitely be noted as something to figure out before stabilization.